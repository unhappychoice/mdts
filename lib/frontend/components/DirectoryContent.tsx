import FolderIcon from '@mui/icons-material/Folder';
import InsertDriveFileIcon from '@mui/icons-material/InsertDriveFile';
import {
  Box,
  Breadcrumbs,
  CircularProgress,
  Divider,
  Link,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  Typography
} from '@mui/material';
import React from 'react';
import { useFileTreeContext } from '../contexts/FileTreeContext';

interface DirectoryContentProps {
  selectedDirectoryPath: string;
  onFileSelect: (filePath: string) => void;
  onDirectorySelect: (directoryPath: string) => void;
}

interface FileTreeItem {
  [key: string]: FileTreeItem[] | string;
}

const filterFileTree = (tree: FileTreeItem[] | string[], targetPath: string): FileTreeItem[] | string[] => {
  if (targetPath === '') {
    const result: FileTreeItem[] | string[] = [];
    tree.forEach(item => {
      if (typeof item === 'string') {
        result.push(item.split('/').pop() || item);
      } else {
        const key = Object.keys(item)[0];
        const newObject: FileTreeItem = {};
        newObject[key.split('/').pop() || key] = item[key];
        result.push(newObject);
      }
    });
    return result;
  }

  const findChildren = (currentTree: FileTreeItem[] | string[], pathSegments: string[], currentSegmentIndex: number): FileTreeItem[] | string[] | null => {
    if (currentSegmentIndex === pathSegments.length) {
      return currentTree;
    }

    const segment = pathSegments[currentSegmentIndex];

    for (const item of currentTree) {
      if (typeof item !== 'string') {
        const key = Object.keys(item)[0];
        const itemSegments = key.split('/');
        if (itemSegments[itemSegments.length - 1] === segment) {
          const children = item[key];
          if (Array.isArray(children)) {
            return findChildren(children, pathSegments, currentSegmentIndex + 1);
          }
        }
      }
    }
    return null;
  };

  const pathSegments = targetPath.split('/').filter(s => s !== '');
  const children = findChildren(tree, pathSegments, 0);

  if (children) {
    const result: FileTreeItem[] | string[] = [];
    children.forEach(item => {
      if (typeof item === 'string') {
        result.push(item.split('/').pop() || item);
      } else {
        const key = Object.keys(item)[0];
        const newObject: FileTreeItem = {};
        newObject[key.split('/').pop() || key] = item[key];
        result.push(newObject);
      }
    });
    return result;
  }

  return [];
};

const DirectoryContent: React.FC<DirectoryContentProps> = ({ selectedDirectoryPath, onFileSelect, onDirectorySelect }) => {
  const { fileTree: fullFileTree, loading, error } = useFileTreeContext();
  const fileTree = filterFileTree(fullFileTree, selectedDirectoryPath);


  const pathSegments = selectedDirectoryPath
    ? selectedDirectoryPath.split('/').filter(segment => segment !== '')
    : [];

  if (error) return <p>Error: {error}</p>;

  const handleItemClick = (itemPath: string, isDirectory: boolean) => {
    if (isDirectory) {
      onDirectorySelect(itemPath);
    } else {
      onFileSelect(itemPath);
    }
  };

  return (
    <Box
      sx={{
        flexGrow: 1,
        p: 4,
        bgcolor: 'background.paper',
        maxWidth: '800px',
        margin: '0 auto',
      }}
    >
      {selectedDirectoryPath && (
        <Breadcrumbs aria-label="breadcrumb" sx={{ mb: 4 }}>
          {pathSegments.map((segment, index) => {
            const isLast = index === pathSegments.length - 1;
            const path = pathSegments.slice(0, index + 1).join('/');
            return isLast ? (
              <Typography key={path} color="text.primary">
                {segment}
              </Typography>
            ) : (
              <Link key={path} color="inherit" href="#" onClick={() => onDirectorySelect && onDirectorySelect(path)}>
                {segment}
              </Link>
            );
          })}
        </Breadcrumbs>
      )}
      <Typography variant="h4" gutterBottom mb={8}>
        Directory: {selectedDirectoryPath}
      </Typography>
      <Divider sx={{ paddingLeft: '24px', marginLeft: '-32px', marginRight: '-32px', borderBottom: 1, borderColor: 'divider' }} />
      {loading && (
        <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }}>
          <CircularProgress />
        </Box>
      )}
      <Box mt={4}>
        {fileTree && (
          <List>
            {fileTree.map((item) => {
              let name: string;
              let itemPath: string;
              let isDirectory: boolean;

              if (typeof item === 'string') {
                // It's a file
                name = item;
                itemPath = selectedDirectoryPath === '' ? item : `${selectedDirectoryPath}/${item}`;
                isDirectory = false;
              } else {
                // It's a directory object { [key: string]: FileTreeItem[] | string; }
                const key = Object.keys(item)[0];
                name = key;
                itemPath = selectedDirectoryPath === '' ? key : `${selectedDirectoryPath}/${key}`;
                isDirectory = true;
              }

              return (
                <ListItem button key={itemPath} onClick={() => handleItemClick(itemPath, isDirectory)}>
                  <ListItemIcon>
                    {isDirectory ? <FolderIcon color="primary" /> : <InsertDriveFileIcon />}
                  </ListItemIcon>
                  <ListItemText primary={name} />
                </ListItem>
              );
            })}
          </List>
        )}
      </Box>
    </Box>
  );
};

export default DirectoryContent;
