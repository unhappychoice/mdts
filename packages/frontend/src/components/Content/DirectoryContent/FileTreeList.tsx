import { ArticleOutlined, FolderOutlined } from '@mui/icons-material';
import { List, ListItem, ListItemIcon, ListItemText } from '@mui/material';
import React from 'react';
import { useSelector } from 'react-redux';
import { FileTreeItem } from '../../../store/slices/fileTreeSlice';
import { RootState } from '../../../store/store';

export interface FileTreeListProps {
  fileTree: (FileTreeItem | { [key: string]: (FileTreeItem | object)[] })[];
  handleItemClick: (itemPath: string, isDirectory: boolean) => void;
}

export const FileTreeList: React.FC<FileTreeListProps> = ({ fileTree, handleItemClick }) => {
  const { currentPath } = useSelector((state: RootState) => state.history);

  return (
    <List sx={{ mr: -2, ml: -2 }}>
      {fileTree.map((item) => {
        const isDirectory = !('path' in item);
        const name = isDirectory ? Object.keys(item)[0] : (item as FileTreeItem).path.split('/').pop();
        const itemPath = currentPath === '' ? name : `${currentPath}/${name}`;

        const handleClick = () => {
          handleItemClick(itemPath, isDirectory);
        };

        return (
          <ListItem key={itemPath} button={true} onClick={handleClick}>
            <ListItemIcon sx={{ minWidth: "38px" }}>
              {isDirectory ? <FolderOutlined color="primary" /> : <ArticleOutlined />}
            </ListItemIcon>
            <ListItemText primary={name} />
          </ListItem>
        );
      })}
    </List>
  );
};
