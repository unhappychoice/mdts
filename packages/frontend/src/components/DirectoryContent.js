"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const icons_material_1 = require("@mui/icons-material");
const material_1 = require("@mui/material");
const FileTreeContext_1 = require("../contexts/FileTreeContext");
const filterFileTree = (tree, targetPath) => {
    if (targetPath === '') {
        const result = [];
        tree.forEach(item => {
            if (typeof item === 'string') {
                result.push(item.split('/').pop() || item);
            }
            else {
                const key = Object.keys(item)[0];
                const newObject = {};
                newObject[key.split('/').pop() || key] = item[key];
                result.push(newObject);
            }
        });
        return result;
    }
    const findChildren = (currentTree, pathSegments, currentSegmentIndex) => {
        if (currentSegmentIndex === pathSegments.length) {
            return currentTree;
        }
        const segment = pathSegments[currentSegmentIndex];
        for (const item of currentTree) {
            if (typeof item !== 'string') {
                const key = Object.keys(item)[0];
                const itemSegments = key.split('/');
                if (itemSegments[itemSegments.length - 1] === segment) {
                    const children = item[key];
                    if (Array.isArray(children)) {
                        return findChildren(children, pathSegments, currentSegmentIndex + 1);
                    }
                }
            }
        }
        return null;
    };
    const pathSegments = targetPath.split('/').filter(s => s !== '');
    const children = findChildren(tree, pathSegments, 0);
    if (children) {
        const result = [];
        children.forEach(item => {
            if (typeof item === 'string') {
                result.push(item.split('/').pop() || item);
            }
            else {
                const key = Object.keys(item)[0];
                const newObject = {};
                newObject[key.split('/').pop() || key] = item[key];
                result.push(newObject);
            }
        });
        return result;
    }
    return [];
};
const DirectoryContent = ({ selectedDirectoryPath, onFileSelect, onDirectorySelect, contentMode = 'fixed' }) => {
    const { fileTree: fullFileTree, loading, error } = (0, FileTreeContext_1.useFileTreeContext)();
    const fileTree = filterFileTree(fullFileTree, selectedDirectoryPath);
    const pathSegments = selectedDirectoryPath
        ? selectedDirectoryPath.split('/').filter(segment => segment !== '')
        : [];
    if (error)
        return (0, jsx_runtime_1.jsxs)("p", { children: ["Error: ", error] });
    const handleItemClick = (itemPath, isDirectory) => {
        if (isDirectory) {
            onDirectorySelect(itemPath);
        }
        else {
            onFileSelect(itemPath);
        }
    };
    return ((0, jsx_runtime_1.jsxs)(material_1.Box, { sx: Object.assign({ flexGrow: 1, p: 4, bgcolor: 'background.paper' }, (contentMode === 'fixed' && {
            maxWidth: '800px',
            margin: '0 auto',
        })), children: [selectedDirectoryPath && ((0, jsx_runtime_1.jsx)(material_1.Breadcrumbs, { "aria-label": "breadcrumb", sx: { mb: 4 }, children: pathSegments.map((segment, index) => {
                    const isLast = index === pathSegments.length - 1;
                    const path = pathSegments.slice(0, index + 1).join('/');
                    return isLast ? ((0, jsx_runtime_1.jsx)(material_1.Typography, { color: "text.primary", children: segment }, path)) : ((0, jsx_runtime_1.jsx)(material_1.Link, { color: "inherit", href: "#", onClick: () => onDirectorySelect && onDirectorySelect(path), children: segment }, path));
                }) })), (0, jsx_runtime_1.jsxs)(material_1.Box, { sx: { display: 'flex', alignItems: 'center', mb: 8 }, children: [(0, jsx_runtime_1.jsx)(icons_material_1.FolderOutlined, { sx: { mr: 2 }, color: "primary", fontSize: "large" }), (0, jsx_runtime_1.jsx)(material_1.Typography, { variant: "h4", gutterBottom: true, sx: { mb: 0 }, children: selectedDirectoryPath })] }), (0, jsx_runtime_1.jsx)(material_1.Divider, { sx: { paddingLeft: '24px', marginLeft: '-32px', marginRight: '-32px', borderBottom: 1, borderColor: 'divider' } }), loading && ((0, jsx_runtime_1.jsx)(material_1.Box, { sx: { display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }, children: (0, jsx_runtime_1.jsx)(material_1.CircularProgress, {}) })), (0, jsx_runtime_1.jsx)(material_1.Box, { mt: 4, children: fileTree && ((0, jsx_runtime_1.jsx)(material_1.List, { sx: { mr: -2, ml: -2 }, children: fileTree.map((item) => {
                        let name;
                        let itemPath;
                        let isDirectory;
                        if (typeof item === 'string') {
                            // It's a file
                            name = item;
                            itemPath = selectedDirectoryPath === '' ? item : `${selectedDirectoryPath}/${item}`;
                            isDirectory = false;
                        }
                        else {
                            // It's a directory object { [key: string]: FileTreeItem[] | string; }
                            const key = Object.keys(item)[0];
                            name = key;
                            itemPath = selectedDirectoryPath === '' ? key : `${selectedDirectoryPath}/${key}`;
                            isDirectory = true;
                        }
                        return ((0, jsx_runtime_1.jsxs)(material_1.ListItem, { button: true, onClick: () => handleItemClick(itemPath, isDirectory), children: [(0, jsx_runtime_1.jsx)(material_1.ListItemIcon, { sx: { minWidth: "38px" }, children: isDirectory ? (0, jsx_runtime_1.jsx)(icons_material_1.FolderOutlined, { color: "primary" }) : (0, jsx_runtime_1.jsx)(icons_material_1.ArticleOutlined, {}) }), (0, jsx_runtime_1.jsx)(material_1.ListItemText, { primary: name })] }, itemPath));
                    }) })) })] }));
};
exports.default = DirectoryContent;
