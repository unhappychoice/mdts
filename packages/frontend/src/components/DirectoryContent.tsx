import { ArticleOutlined, FolderOutlined } from '@mui/icons-material';
import {
  Box,
  Breadcrumbs,
  CircularProgress,
  Divider,
  Link,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  Typography
} from '@mui/material';
import React from 'react';
import { useSelector } from 'react-redux';
import { RootState } from '../store/store';
import { selectFilteredFileTree } from '../store/slices/fileTreeSlice';

interface DirectoryContentProps {
  selectedDirectoryPath: string;
  onFileSelect: (filePath: string) => void;
  onDirectorySelect: (directoryPath: string) => void;
  contentMode?: 'fixed' | 'full';
}

const DirectoryContent: React.FC<DirectoryContentProps> = ({ selectedDirectoryPath, onFileSelect, onDirectorySelect, contentMode = 'fixed' }) => {
  const { fileTree: fullFileTree, loading, error } = useSelector((state: RootState) => state.fileTree);
  const fileTree = selectFilteredFileTree(fullFileTree, selectedDirectoryPath);

  const pathSegments = selectedDirectoryPath
    ? selectedDirectoryPath.split('/').filter(segment => segment !== '')
    : [];

  if (error) return <p>Error: {error}</p>;

  const handleItemClick = (itemPath: string, isDirectory: boolean) => {
    if (isDirectory) {
      onDirectorySelect(itemPath);
    } else {
      onFileSelect(itemPath);
    }
  };

  return (
    <Box
      sx={{
        flexGrow: 1,
        p: 4,
        bgcolor: 'background.paper',
        ...(contentMode === 'fixed' && {
          maxWidth: '800px',
          margin: '0 auto',
        }),
      }}
    >
      {selectedDirectoryPath && (
        <Breadcrumbs aria-label="breadcrumb" sx={{ mb: 4 }}>
          {pathSegments.map((segment, index) => {
            const isLast = index === pathSegments.length - 1;
            const path = pathSegments.slice(0, index + 1).join('/');
            return isLast ? (
              <Typography key={path} color="text.primary">
                {segment}
              </Typography>
            ) : (
              <Link key={path} color="inherit" href="#" onClick={() => onDirectorySelect && onDirectorySelect(path)}>
                {segment}
              </Link>
            );
          })}
        </Breadcrumbs>
      )}
      <Box sx={{ display: 'flex', alignItems: 'center', mb: 8 }}>
        <FolderOutlined sx={{ mr: 2 }} color="primary" fontSize="large"/>
        <Typography variant="h4" gutterBottom sx={{ mb: 0 }}>
          {selectedDirectoryPath}
        </Typography>
      </Box>
      <Divider sx={{ paddingLeft: '24px', marginLeft: '-32px', marginRight: '-32px', borderBottom: 1, borderColor: 'divider' }} />
      {loading && (
        <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }}>
          <CircularProgress />
        </Box>
      )}
      <Box mt={4}>
        {fileTree && (
          <List sx={{ mr: -2, ml: -2 }}>
            {fileTree.map((item) => {
              let name: string;
              let itemPath: string;
              let isDirectory: boolean;

              if (typeof item === 'string') {
                // It's a file
                name = item;
                itemPath = selectedDirectoryPath === '' ? item : `${selectedDirectoryPath}/${item}`;
                isDirectory = false;
              } else {
                // It's a directory object { [key: string]: FileTreeItem[] | string; }
                const key = Object.keys(item)[0];
                name = key;
                itemPath = selectedDirectoryPath === '' ? key : `${selectedDirectoryPath}/${key}`;
                isDirectory = true;
              }

              return (
                <ListItem button key={itemPath} onClick={() => handleItemClick(itemPath, isDirectory)}>
                  <ListItemIcon sx={{ minWidth: "38px" }}>
                    {isDirectory ? <FolderOutlined color="primary" /> : <ArticleOutlined />}
                  </ListItemIcon>
                  <ListItemText primary={name} />
                </ListItem>
              );
            })}
          </List>
        )}
      </Box>
    </Box>
  );
};

export default DirectoryContent;
