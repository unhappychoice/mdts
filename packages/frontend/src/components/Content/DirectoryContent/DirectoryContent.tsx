import { FolderOutlined } from '@mui/icons-material';
import { Box, CircularProgress, Divider, Typography } from '@mui/material';
import React, { useCallback } from 'react';
import { useSelector } from 'react-redux';
import { selectFilteredFileTree } from '../../../store/slices/fileTreeSlice';
import { RootState } from '../../../store/store';
import ErrorView from '../../ErrorView';
import BreadCrumb from '../BreadCrumb';
import { FileTreeList } from './FileTreeList';

interface DirectoryContentProps {
  onFileSelect: (filePath: string) => void;
  onDirectorySelect: (directoryPath: string) => void;
}

const DirectoryContent: React.FC<DirectoryContentProps> = ({ onFileSelect, onDirectorySelect }) => {
  const { currentPath } = useSelector((state: RootState) => state.history);
  const { contentMode } = useSelector((state: RootState) => state.appSetting);
  const { fileTree: fullFileTree, loading, error } = useSelector((state: RootState) => state.fileTree);

  const fileTree = selectFilteredFileTree(fullFileTree, currentPath);

  const handleItemClick = useCallback((itemPath: string, isDirectory: boolean) => {
    if (isDirectory) {
      onDirectorySelect(itemPath);
    } else {
      onFileSelect(itemPath);
    }
  }, [onDirectorySelect, onFileSelect]);

  if (error) {
    return <ErrorView error={error} />;
  }

  return (
    <Box
      sx={{
        width: '100%',
        minHeight: 'calc(100vh - 64px)',
        m: 0,
        p: 4,
        bgcolor: 'background.paper',
        ...(contentMode === 'compact' && {
          width: '800px',
          margin: '0 auto',
        })
      }}
    >
      <BreadCrumb onDirectorySelect={onDirectorySelect} />
      <Box sx={{ display: 'flex', alignItems: 'center', mb: 4 }}>
        <FolderOutlined sx={{ mr: 2 }} color="primary" fontSize="large" />
        <Typography variant="h4" gutterBottom sx={{ mb: 0 }}>
          {currentPath}
        </Typography>
      </Box>
      <Divider sx={{ paddingLeft: '24px', marginLeft: '-32px', marginRight: '-32px', borderBottom: 1, borderColor: 'divider' }} />
      {loading && (
        <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }}>
          <CircularProgress />
        </Box>
      )}
      <Box mt={4}>
        <FileTreeList fileTree={fileTree} handleItemClick={handleItemClick} />
      </Box>
    </Box>
  );
};

export default DirectoryContent;
