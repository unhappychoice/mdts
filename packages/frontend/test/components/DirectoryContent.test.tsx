import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import DirectoryContent from '../../src/components/DirectoryContent';

const mockStore = configureStore([]);

describe('DirectoryContent', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      fileTree: {
        fileTree: [{ name: 'file1.md', path: '/file1.md', type: 'file' }],
        filteredFileTree: [{ name: 'file1.md', path: '/file1.md', type: 'file' }],
        loading: false,
        error: null,
        searchQuery: '',
      },
    });
  });

  test('renders correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <DirectoryContent selectedDirectoryPath="" onFileSelect={() => {}} onDirectorySelect={() => {}} contentMode="fixed" />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
