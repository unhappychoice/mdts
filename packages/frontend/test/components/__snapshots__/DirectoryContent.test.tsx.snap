// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`DirectoryContent renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-cf17b1"
  >
    <div
      class="MuiBox-root css-1v3fpe7"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeLarge css-1nzojzr-MuiSvgIcon-root"
        data-testid="FolderOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m9.17 6 2 2H20v10H4V6zM10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8z"
        />
      </svg>
      <h4
        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-129bu66-MuiTypography-root"
      />
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-1aavb8m-MuiDivider-root"
    />
    <div
      class="MuiBox-root css-h5fkc8"
    >
      <ul
        class="MuiList-root MuiList-padding css-1tfyysn-MuiList-root"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-z2qibm-MuiListItem-root"
        >
          <div
            class="MuiListItemIcon-root css-1sms2q2-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-1wtp6o0-MuiSvgIcon-root"
              data-testid="FolderOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m9.17 6 2 2H20v10H4V6zM10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-1px98du-MuiTypography-root"
            >
              name
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
